tests/python/pants_test/backend/codegen/tasks/test_simple_codegen_task.py

```diff
@@ -0,0 +1,230 @@
+    isolated_dirs = set(get_suffix(target, 'isolated') for target in targets)
+
+    self.assertEqual(1, len(global_dirs), 'There should only be one global directory suffix!')
+    self.assertEqual(len(targets), len(isolated_dirs),
+                     'There should be exactly one directory suffix per unique target!')
+
+    for target in targets:
+      for strategy in ('global', 'isolated',):
+        suffix = None
+        for i in range(5):
+          curr_suffix = get_suffix(target, strategy)
+          if i > 0:
+            self.assertEqual(suffix, curr_suffix,
+                             'Suffix should not be unstable for the same target and strategy! '
+                             '({target} with {strategy} got {suffix_a} and {suffix_b})'
+                             .format(target=target.id,
+                                     strategy=strategy,
+                                     suffix_a=suffix,
+                                     suffix_b=curr_suffix))
+          suffix = curr_suffix
+
+  def test_execute(self):
+    dummy_suffixes = ['a', 'b', 'c',]
+
+    self.add_to_build_file('gen-lib', '\n'.join(dedent('''
+      dummy_library(name='{suffix}',
+        sources=['org/pantsbuild/example/foo{suffix}.dummy'],
+      )
+    ''').format(suffix=suffix) for suffix in dummy_suffixes))
+
+    for suffix in dummy_suffixes:
+      self.create_file('gen-lib/org/pantsbuild/example/foo{suffix}.dummy'.format(suffix=suffix),
+                       'org.pantsbuild.example Foo{0}'.format(suffix))
+
+    targets = [self.target('gen-lib:{suffix}'.format(suffix=suffix)) for suffix in dummy_suffixes]
+    for strategy in ('global', 'isolated',):
+      task = self._create_dummy_task(target_roots=targets, strategy=strategy)
+      expected_targets = set(targets)
+      found_targets = set(task.codegen_targets())
+      self.assertEqual(expected_targets, found_targets,
+                       'TestGen failed to find codegen target {expected}! Found: [{found}].'
+                       .format(expected=', '.join(t.id for t in expected_targets),
+                               found=', '.join(t.id for t in found_targets)))
+      task.execute()
+
+
+  class DummyLibrary(JvmTarget):
+    """Library of .dummy files, which are just text files which generate empty java files.
+
+    As the name implies, this is purely for testing the behavior of the simple_codegen_task.
+
+    For example, a .dummy file with the contents:
+    org.company.package Main
+    org.company.package Foobar
+    org.company.other Barfoo
+
+    Would generate the files:
+    org/company/package/Main.java,
+    org/company/package/Foobar.java,
+    org/company/other/Barfoo.java,
+
+    Which would compile, but do nothing.
+    """
+
+    def __init__(self, payload=None, **kwargs):
+      payload = payload or Payload()
+      super(SimpleCodegenTaskTest.DummyLibrary, self).__init__(payload=payload, **kwargs)
+      self.add_labels('codegen')
```

Comment:
Note that we are trying to expunge the use of labels in the code.  I'm looking and I don't think we actually need this label set for this test, but it probably won't hurt anything.